hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/HiveTypeToAvroType.java

```diff
@@ -0,0 +1,272 @@
+import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector;
+import org.apache.hadoop.hive.serde2.typeinfo.DecimalTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.ListTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.MapTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.PrimitiveTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.UnionTypeInfo;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.codehaus.jackson.node.JsonNodeFactory;
+
+
+public class HiveTypeToAvroType {
+  private int recordCounter;
+  private final String namespace;
+
+  private final boolean mkFieldsOptional;
+
+  // Additional numeric type, similar to other logical type names in AvroSerde
+  private static final String SHORT_TYPE_NAME = "short";
+  private static final String BYTE_TYPE_NAME = "byte";
+
+  public HiveTypeToAvroType(String namespace, boolean mkFieldsOptional) {
+    this.recordCounter = 0;
+    this.namespace = namespace;
+
+    this.mkFieldsOptional = mkFieldsOptional;
+  }
+
+  Schema convertFieldsTypeInfoToAvroSchema(String recordNamespace, String recordName, List<String> fieldNames,
+      List<TypeInfo> fieldTypeInfos) {
+    final List<Schema.Field> fields = new ArrayList<>();
+    for (int i = 0; i < fieldNames.size(); ++i) {
+      final TypeInfo fieldTypeInfo = fieldTypeInfos.get(i);
+      String fieldName = fieldNames.get(i);
+      fieldName = removePrefix(fieldName);
+
+      // If there's a structType in the schema, we will use "recordNamespace.fieldName" instead of the
+      // autogenerated record name. The recordNamespace is composed of its parent's field names recursively.
+      // This mimics the logic of spark-avro.
+      // We will set the recordName to be capitalized, and the recordNameSpace will be in lower case
+      final Schema schema = convertTypeInfoToAvroSchema(fieldTypeInfo, recordNamespace + "." + recordName.toLowerCase(),
+          StringUtils.capitalize(fieldName));
+      final Schema.Field f = new Schema.Field(fieldName, schema, null, null);
+      fields.add(f);
+    }
+
+    final Schema recordSchema = Schema.createRecord(recordName, null, namespace + recordNamespace, false);
+    recordSchema.setFields(fields);
+    return recordSchema;
+  }
+
+  Schema convertTypeInfoToAvroSchema(TypeInfo typeInfo, String recordNamespace, String recordName) {
+    Schema schema;
+    ObjectInspector.Category category = typeInfo.getCategory();
+
+    switch (category) {
+      case STRUCT:
+        // We don't cache the structType because otherwise it could be possible that a field
+        // "lastname" is of type "firstname", where firstname is a compiled class.
+        // This will lead to ambiguity.
+        schema = parseSchemaFromStruct((StructTypeInfo) typeInfo, recordNamespace, recordName);
```

Comment:
Can we rename this and other methods to something like `convertStructTypeInfoToAvroSchema`?